engine/src/main/java/org/terasology/physics/shapes/CollisionShapeFactory.java

```diff
@@ -62,4 +64,13 @@ default BoxShape getNewUnitCube() {
      * @return The created sphere shape.
      */
     SphereShape getNewSphere(float radius);
+
+
+    /**
+     * Returns the colliding body associated with the entity
+     *
+     * @param entityRef
+     * @return
+     */
+    BulletCollisionShape getShapeFor(EntityRef entityRef);
```

Comment:
Since a `Factory` is responsible for creating objects, I would prefer moving this method somewhere else. Besides, this method does not need to be overridable -- similar to your implementation in `BulletCollisionShapeFactory`, the final implementation will look for components on the entity and return shapes created from a `CollisionShapeFactory` instance. Also, in the interest of remaining physics engine agnostic, the return type should be made `CollisionShape`. Sample final implementation:

```java
CollisionShape getShapeFor(EntityRef entity, CollisionShapeFactory collisionShapeFactory) {
        BoxShapeComponent box = entityRef.getComponent(BoxShapeComponent.class);
        if (box != null) {
            Vector3f halfExtents = new Vector3f(box.extents);
            return collisionShapeFactory.getNewBox(halfExtents.mul(.5f));
        }
        SphereShapeComponent sphere = entityRef.getComponent(SphereShapeComponent.class);
        if (sphere != null) {
            return collisionShapeFactory.getNewSphere(sphere.radius);
        }
        CapsuleShapeComponent capsule = entityRef.getComponent(CapsuleShapeComponent.class);
        if (capsule != null) {
            // This method does not yet exist in CollisionShapeFactory
            return collisionShapeFactory.getNewCapsule(capsule.radius,capsule.height);
        }
        CylinderShapeComponent cylinder = entityRef.getComponent(CylinderShapeComponent.class);
        if (cylinder != null) {
            // This method does not yet exist in CollisionShapeFactory
            return collisionShapeFactory.getNewCylinder(new Vector3f(cylinder.radius, 0.5f * cylinder.height, cylinder.radius));
        }
        HullShapeComponent hull = entityRef.getComponent(HullShapeComponent.class);
        if (hull != null) {
            FloatBuffer buffer = BufferUtils.createFloatBuffer( hull.sourceMesh.getVertices().size());
            TFloatIterator iterator = hull.sourceMesh.getVertices().iterator();
            int numPoints = 0;
            while (iterator.hasNext()) {
                numPoints++;
                buffer.put(iterator.next());
                buffer.put(iterator.next());
                buffer.put(iterator.next());
            }
            return collisionShapeFactory.getNewConvexHull(buffer,numPoints,3 * Float.BYTES);
        }
        CharacterMovementComponent characterMovementComponent = entityRef.getComponent(CharacterMovementComponent.class);
        if (characterMovementComponent != null) {
            return collisionShapeFactory.getNewCapsule(characterMovementComponent.radius,characterMovementComponent.height);
        }
        throw new IllegalArgumentException("Creating physics object that requires a ShapeComponent or CharacterMovementComponent, but has neither. Entity: " + entityRef);
    }
```